src/gcp/cloudsql/permissions_setup.ts

```diff
@@ -0,0 +1,384 @@
+    BEGIN
+        -- Count the number of rows matching the criteria
+        SELECT COUNT(*)
+        INTO role_count
+        FROM
+          pg_auth_members m
+        JOIN
+          pg_roles grantee ON grantee.oid = m.member
+        JOIN
+          pg_roles granted ON granted.oid = m.roleid
+        JOIN
+          pg_roles grantor ON grantor.oid = m.grantor
+        WHERE
+          granted.rolname = '${grantedRole}'
+          AND grantee.rolname = '${granteeRole}';
+
+        -- If no rows were found, raise an exception
+        IF role_count = 0 THEN
+            RAISE EXCEPTION 'Role "%", is not granted to role "%".', '${grantedRole}', '${granteeRole}';
+        END IF;
+    END $$;
+`;
+  try {
+    await executeSqlCmdsAsIamUser(options, instanceId, databaseId, [checkCmd], /** silent=*/ true);
+    return true;
+  } catch (e) {
+    // We only return false after we confirm the error is indeed because the role isn't granted.
+    // Otherwise we propagate the error.
+    if (e instanceof FirebaseError && e.message.includes("not granted to role")) {
+      return false;
+    }
+    logger.error(`Role Check Failed: ${e}`);
+    throw e;
+  }
+}
+
+// Sets up all FDC roles (owner, writer, and reader).
+// Granting roles to users is done by the caller.
+// Returns true if schema is setup as greenfield
+export async function setupSQLPermissions(
+  instanceId: string,
+  databaseId: string,
+  schemaInfo: SchemaMetaData,
+  options: Options,
+  silent: boolean = false,
+): Promise<SchemaSetupStatus.BrownField | SchemaSetupStatus.GreenField> {
+  const schema = schemaInfo.name;
+  // Step 0: Check current user can run setup and upsert IAM / P4SA users
+  logger.info(`Attempting to Setup SQL schema "${schema}".`);
+  const userIsCSQLAdmin = await iamUserIsCSQLAdmin(options);
+  if (!userIsCSQLAdmin) {
+    throw new FirebaseError(
+      `Missing required IAM permission to setup SQL schemas. SQL schema setup requires 'roles/cloudsql.admin' or an equivalent role.`,
+    );
+  }
+  await setupIAMUsers(instanceId, databaseId, options);
+
+  if (schemaInfo.setupStatus === SchemaSetupStatus.GreenField) {
+    logger.info(`Seems like the database is already setup. Reruning the setup process.`);
+    await greenFieldSchemaSetup(instanceId, databaseId, schema, options, silent);
+    return SchemaSetupStatus.GreenField;
+  } else {
+    logger.info(`Detected schema "${schema}" setup status is ${schemaInfo.setupStatus}.`);
+  }
+
+  // We need to setup the database
+  if (schemaInfo.tables.length === 0) {
+    logger.info(`Found no tables in schema "${schema}", assuming greenfield project.`);
+    await greenFieldSchemaSetup(instanceId, databaseId, schema, options, silent);
+    logger.info(clc.green("Database setup complete."));
+    return SchemaSetupStatus.GreenField;
+  }
+
+  if (options.nonInteractive || options.force) {
+    throw new FirebaseError(
+      `Schema "${schema}" isn't setup and can only be setup in interactive mode.`,
```

Comment:
```suggestion
      `Schema "${schema}" isn't set up and can only be set up in interactive mode.`,
```
Had to look this one up to confirm, but generally setup is a noun, set up is a verb: https://grammarist.com/spelling/set-up-vs-setup/